std/digest/murmurhash.d

```diff
@@ -0,0 +1,850 @@
+    this(uint seed)
+    {
+        h1 = seed;
+    }
+
+    @disable this(this);
+
+    /// Adds a single Block of data without increasing size.
+    /// Make sure to increase size by Block.sizeof for each call to putBlock.
+    void putBlock(uint block) pure nothrow @nogc
+    {
+        h1 = update(h1, block, 0, c1, c2, 15, 13, 0xe6546b64U);
+    }
+
+    /// Put remainder bytes. This must be called only once after putBlock and before finalize.
+    void putRemainder(scope const(ubyte[]) data...) pure nothrow @nogc
+    {
+        assert(data.length < Block.sizeof);
+        assert(data.length >= 0);
+        size += data.length;
+        uint k1 = 0;
+        final switch (data.length & 3)
+        {
+        case 3:
+            k1 ^= data[2] << 16;
+            goto case;
+        case 2:
+            k1 ^= data[1] << 8;
+            goto case;
+        case 1:
+            k1 ^= data[0];
+            h1 ^= shuffle(k1, c1, c2, 15);
+            goto case;
+        case 0:
+        }
+    }
+
+    /// Incorporate size and finalizes the hash.
+    void finalize() pure nothrow @nogc
+    {
+        h1 ^= size;
+        h1 = fmix(h1);
+    }
+
+    /// Returns the hash as an uint value.
+    Block get() pure nothrow @nogc
+    {
+        return h1;
+    }
+
+    /**
+    Pushes an array of blocks at once. It is more efficient to push as much data as possible in a single call.
+    On platform that does not support unaligned reads (MIPS or old ARM chips), the compiler may produce slower code to ensure correctness.
+    */
+    void putBlocks(scope const(Block[]) blocks...) pure nothrow @nogc
+    {
+        foreach (const block; blocks)
+        {
+            putBlock(block);
+        }
+        size += blocks.length * Block.sizeof;
+    }
+
+    /// Returns the current hashed value as an ubyte array.
+    auto getBytes() pure nothrow @nogc
+    {
+        return cast(ubyte[Block.sizeof]) cast(uint[1])[get()];
+    }
+}
+
+version (unittest)
+{
+    import std.string : representation;
+
+    auto hash(H, Block = H.Block)(string data)
+    {
+        H hasher;
+        immutable blocks = data.length / Block.sizeof;
+        hasher.putBlocks(cast(const(Block)[]) data[0 .. blocks * Block.sizeof]);
+        hasher.putRemainder(cast(const(ubyte)[]) data[blocks * Block.sizeof .. $]);
+        hasher.finalize();
+        return hasher.getBytes();
+    }
+
+    void checkResult(H)(in string[string] groundtruth)
+    {
+        foreach (data, expectedHash; groundtruth)
+        {
+            alias PiecewiseH = Piecewise!H;
+            assert(data.digest!PiecewiseH.toHexString() == expectedHash);
+            assert(data.hash!H.toHexString() == expectedHash);
+            PiecewiseH hasher;
+            foreach (element; data)
+            {
+                hasher.put(element);
+            }
+            assert(hasher.finish.toHexString() == expectedHash);
+        }
+    }
+}
+
+unittest
+{
+    // dfmt off
+    checkResult!FastMurmurHash3_x86_32([
```

Comment:
If you're going to add verification values, take a look at how SMHasher does things in VerificationTest() -

https://github.com/aappleby/smhasher/blob/master/src/KeysetTest.cpp
https://github.com/aappleby/smhasher/blob/master/src/main.cpp